src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -202,33 +226,102 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
-            ExceptionDispatchInfo lastException = null;
+            var sockets = new List<Socket>(serverAddresses.Length);
+            var connectTasks = new List<Task>(serverAddresses.Length);
+            var tcs = new TaskCompletionSource<Socket>();
+            var lastError = new StrongBox<Exception>();
+            var pendingCompleteCount = new StrongBox<int>(serverAddresses.Length);
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                var socket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                sockets.Add(socket);
+
+                // Start all connection tasks now, to prevent possible race conditions with
+                // calling ConnectAsync on disposed sockets.
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    connectTasks.Add(socket.ConnectAsync(address, port));
                 }
-                catch (Exception exc)
+                catch (Exception e)
                 {
-                    socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    connectTasks.Add(Task.FromException(e));
                 }
             }
 
-            // Propagate the last failure that occurred
-            if (lastException != null)
+            for (int i = 0; i < sockets.Count; i++)
             {
-                lastException.Throw();
+                ParallelConnectHelper(sockets[i], connectTasks[i], tcs, pendingCompleteCount, lastError, sockets);
             }
 
-            // Should never get here.  Either there will have been no addresses and we'll have thrown
-            // at the beginning, or one of the addresses will have worked and we'll have returned, or
-            // at least one of the addresses will failed, in which case we will have propagated that.
-            throw new ArgumentException();
+            return tcs.Task;
+        }
+
+        private static async void ParallelConnectHelper(
+            Socket socket,
+            Task connectTask,
+            TaskCompletionSource<Socket> tcs,
+            StrongBox<int> pendingCompleteCount,
+            StrongBox<Exception> lastError,
+            List<Socket> sockets)
+        {
+            bool success = false;
+            try
+            {
+                // Try to connect.  If we're successful, store this task into the result task.
+                await connectTask.ConfigureAwait(false);
+                if (connectTask.IsFaulted)
+                {
+                    Interlocked.Exchange(ref lastError.Value, connectTask.Exception);
+                }
+                else if (socket.Connected)
```

Comment:
I'm not sure this check is necessary.  If we get here, the ConnectAsync task completed successfully.  In the other functions, e.g. https://github.com/dotnet/corefx/pull/15494/files#diff-f8811317213fc5caaab5bf7d43a68000R178, we assume that if the ConnectAsync task completes successfully, then we have a valid socket to use.  I'm not sure why this case is different.